js_modules/dagit/packages/core/src/app/CodeLinkProtocol.tsx

```diff
@@ -0,0 +1,94 @@
+import {MenuItem, Menu, Select, TextInput, Box, Button, Icon, CaptionMono} from '@dagster-io/ui';
+import * as React from 'react';
+
+import {useStateWithStorage} from '../hooks/useStateWithStorage';
+
+export const CodeLinkProtocolKey = 'CodeLinkProtocolPreference';
+
+const POPULAR_PROTOCOLS = {
+  'vscode://file/{FILE}:{LINE}': 'Visual Studio Code',
+  '': 'Custom',
+};
+
+const DEFAULT_PROTOCOL = {protocol: Object.keys(POPULAR_PROTOCOLS)[0], custom: false};
+
+type ProtocolData = {
+  protocol: string;
+  custom: boolean;
+};
+
+export const CodeLinkProtocolContext = React.createContext<
+  [ProtocolData, React.Dispatch<React.SetStateAction<ProtocolData | undefined>>]
+>([DEFAULT_PROTOCOL, () => '']);
+
+export const CodeLinkProtocolProvider: React.FC = (props) => {
+  const state = useStateWithStorage<ProtocolData>(CodeLinkProtocolKey, (x) =>
+    x === undefined ? DEFAULT_PROTOCOL : x,
+  );
+
```

Comment:
This looks good to me, but I think you could actually write this combo of a context + useStateWithStorage as a custom hook and that might be a bit less code.

Multiple references to `useStateWithStorage(key)` with the same `key` are kept in sync on the page, so it'd be safe to write this like this. If two components are both using your hook, one of them writing a new value will cause the other to update even without a custom context :-)  

```
const useCodeLinkProtocol() {
     return useStateWithStorage<ProtocolData>(CodeLinkProtocolKey, (x) =>
    x ?? DEFAULT_PROTOCOL);
}

```